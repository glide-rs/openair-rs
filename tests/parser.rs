use std::io::Cursor;

use indoc::indoc;
use insta::assert_debug_snapshot;
use openair::*;

#[test]
fn test_switzerland_fixture() {
    let airspace = include_bytes!("../example_data/Switzerland.txt");
    let mut cursor = Cursor::new(airspace);

    let spaces = parse(&mut cursor).unwrap();
    assert_debug_snapshot!(spaces.first().unwrap());

    let names = spaces.iter().map(|a| a.to_string()).collect::<Vec<_>>();
    assert_debug_snapshot!(names);
}

#[test]
fn test_germany_fixture() {
    let airspace = include_bytes!("../example_data/Germany.txt");
    let mut cursor = Cursor::new(airspace);

    let spaces = parse(&mut cursor).unwrap();
    assert_debug_snapshot!(spaces.first().unwrap());

    let names = spaces.iter().map(|a| a.to_string()).collect::<Vec<_>>();
    assert_debug_snapshot!(names);
}

#[test]
fn test_germany_border_fixture() {
    let airspace = include_bytes!("../example_data/Germany_Border.txt");
    let mut cursor = Cursor::new(airspace);

    let spaces = parse(&mut cursor).unwrap();
    assert_debug_snapshot!(spaces.first().unwrap());

    let names = spaces.iter().map(|a| a.to_string()).collect::<Vec<_>>();
    assert_debug_snapshot!(names);
}

/// Parse an airspace as generated by flyland.
#[test]
fn flyland_buochs() {
    let mut airspace = indoc! {"
        AC D
        AN BUOCHS Be CTR 119.625
        AL GND
        AH 12959 ft
        DP 46:57:13 N 008:27:52 E
        DP 46:57:46 N 008:30:41 E
        DP 46:57:55 N 008:28:40 E
        DP 46:58:28 N 008:27:56 E
        DP 46:57:13 N 008:27:52 E
        * n-Points: 5
    "}
    .as_bytes();
    let mut spaces = parse(&mut airspace).unwrap();
    assert_eq!(spaces.len(), 1);
    let space: Airspace = spaces.pop().unwrap();
    assert_eq!(space.name, "BUOCHS Be CTR 119.625");
    assert_eq!(space.lower_bound, Altitude::Gnd);
    assert_eq!(space.upper_bound, Altitude::FeetAmsl(12959));
    if let Geometry::Polygon { segments } = space.geom {
        assert_eq!(segments.len(), 5);
    } else {
        panic!("Unexpected enum variant");
    }
}

/// The order of bounds should not matter.
#[test]
fn inverted_bounds() {
    let mut a1 = indoc! {"
        AC D
        AN SOMESPACE
        AL GND
        AH 12959 ft
        DP 46:57:13 N 008:27:52 E
        DP 46:57:46 N 008:30:41 E
        *
    "}
    .as_bytes();
    let mut a2 = indoc! {"
        AC D
        AN SOMESPACE
        AH 12959 ft
        AL GND
        DP 46:57:13 N 008:27:52 E
        DP 46:57:46 N 008:30:41 E
        *
    "}
    .as_bytes();
    let space1 = parse(&mut a1).unwrap().pop().unwrap();
    let space2 = parse(&mut a2).unwrap().pop().unwrap();
    assert_eq!(space1, space2);
}

/// Variables can be defined multiple times.
#[test]
fn multi_variable() {
    let mut a = indoc! {"
        AC D
        AN SOMESPACE
        AL GND
        AH FL100
        V X=52:00:00N 013:00:00E
        V D=+
        DA 2,0,30
        V X=52:00:00N 013:00:00E
        V D=-
        DA 4,60,30
        *
    "}
    .as_bytes();
    let airspace = parse(&mut a).unwrap().pop().unwrap();
    assert_eq!(
        airspace.geom,
        Geometry::Polygon {
            segments: vec![
                PolygonSegment::ArcSegment(ArcSegment {
                    centerpoint: Coord {
                        lat: 52.0,
                        lng: 13.0
                    },
                    radius: 2.0,
                    angle_start: 0.0,
                    angle_end: 30.0,
                    direction: Direction::Cw,
                }),
                PolygonSegment::ArcSegment(ArcSegment {
                    centerpoint: Coord {
                        lat: 52.0,
                        lng: 13.0
                    },
                    radius: 4.0,
                    angle_start: 60.0,
                    angle_end: 30.0,
                    direction: Direction::Ccw,
                }),
            ],
        }
    );
}

/// Test AY/AF/AG records.
#[test]
fn extension_records() {
    let mut a = indoc! {"
        AC D
        AN SOMESPACE
        AL GND
        AH 100 ft AGL
        AY AWY
        AF 132.350
        AG Dutch Mil
        AX 1234
        A* custom extension
        V X=52:00:00 N 013:00:00 E
        DC 5
    "}
    .as_bytes();
    let airspace = parse(&mut a).unwrap().pop().unwrap();
    assert_eq!(airspace.type_, Some("AWY".to_string()));
    assert_eq!(airspace.frequency, Some("132.350".to_string()));
    assert_eq!(airspace.call_sign, Some("Dutch Mil".to_string()));
    assert_eq!(airspace.transponder_code, Some(1234));
}
