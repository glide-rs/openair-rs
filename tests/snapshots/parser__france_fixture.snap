---
source: tests/parser.rs
expression: spaces.first().unwrap()
---
Airspace {
    name: "ZSM GYPAETE BARBU NID FELCE 2025",
    class: Unclassified,
    type_: Some(
        "OFR",
    ),
    lower_bound: Gnd,
    upper_bound: FeetAmsl(
        5577,
    ),
    geom: Polygon {
        segments: [
            Point(
                Coord {
                    lat: 42.27027777777778,
                    lng: 8.821388888888889,
                },
            ),
            Point(
                Coord {
                    lat: 42.271388888888886,
                    lng: 8.821666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 42.2725,
                    lng: 8.821388888888889,
                },
            ),
            Point(
                Coord {
                    lat: 42.27333333333333,
                    lng: 8.821111111111112,
                },
            ),
            Point(
                Coord {
                    lat: 42.27388888888889,
                    lng: 8.820833333333333,
                },
            ),
            Point(
                Coord {
                    lat: 42.27638888888889,
                    lng: 8.818333333333333,
                },
            ),
            Point(
                Coord {
                    lat: 42.277499999999996,
                    lng: 8.818333333333333,
                },
            ),
            Point(
                Coord {
                    lat: 42.27861111111111,
                    lng: 8.818055555555555,
                },
            ),
            Point(
                Coord {
                    lat: 42.27916666666667,
                    lng: 8.818055555555555,
                },
            ),
            Point(
                Coord {
                    lat: 42.27972222222222,
                    lng: 8.817499999999999,
                },
            ),
            Point(
                Coord {
                    lat: 42.27972222222222,
                    lng: 8.81638888888889,
                },
            ),
            Point(
                Coord {
                    lat: 42.27916666666667,
                    lng: 8.815000000000001,
                },
            ),
            Point(
                Coord {
                    lat: 42.27916666666667,
                    lng: 8.812222222222223,
                },
            ),
            Point(
                Coord {
                    lat: 42.27972222222222,
                    lng: 8.81,
                },
            ),
            Point(
                Coord {
                    lat: 42.28,
                    lng: 8.809166666666668,
                },
            ),
            Point(
                Coord {
                    lat: 42.27972222222222,
                    lng: 8.808333333333334,
                },
            ),
            Point(
                Coord {
                    lat: 42.279444444444444,
                    lng: 8.806944444444445,
                },
            ),
            Point(
                Coord {
                    lat: 42.27972222222222,
                    lng: 8.806111111111111,
                },
            ),
            Point(
                Coord {
                    lat: 42.279444444444444,
                    lng: 8.805000000000001,
                },
            ),
            Point(
                Coord {
                    lat: 42.279444444444444,
                    lng: 8.80388888888889,
                },
            ),
            Point(
                Coord {
                    lat: 42.27972222222222,
                    lng: 8.803333333333335,
                },
            ),
            Point(
                Coord {
                    lat: 42.280277777777776,
                    lng: 8.803611111111111,
                },
            ),
            Point(
                Coord {
                    lat: 42.28055555555555,
                    lng: 8.803333333333335,
                },
            ),
            Point(
                Coord {
                    lat: 42.27972222222222,
                    lng: 8.800833333333333,
                },
            ),
            Point(
                Coord {
                    lat: 42.27916666666667,
                    lng: 8.8,
                },
            ),
            Point(
                Coord {
                    lat: 42.27861111111111,
                    lng: 8.799722222222222,
                },
            ),
            Point(
                Coord {
                    lat: 42.27833333333333,
                    lng: 8.799166666666666,
                },
            ),
            Point(
                Coord {
                    lat: 42.27777777777778,
                    lng: 8.797777777777778,
                },
            ),
            Point(
                Coord {
                    lat: 42.27777777777778,
                    lng: 8.796944444444444,
                },
            ),
            Point(
                Coord {
                    lat: 42.278055555555554,
                    lng: 8.795833333333333,
                },
            ),
            Point(
                Coord {
                    lat: 42.27777777777778,
                    lng: 8.795277777777777,
                },
            ),
            Point(
                Coord {
                    lat: 42.27777777777778,
                    lng: 8.794444444444444,
                },
            ),
            Point(
                Coord {
                    lat: 42.27722222222222,
                    lng: 8.793333333333333,
                },
            ),
            Point(
                Coord {
                    lat: 42.276666666666664,
                    lng: 8.7925,
                },
            ),
            Point(
                Coord {
                    lat: 42.275555555555556,
                    lng: 8.791666666666666,
                },
            ),
            Point(
                Coord {
                    lat: 42.274166666666666,
                    lng: 8.790833333333333,
                },
            ),
            Point(
                Coord {
                    lat: 42.27388888888889,
                    lng: 8.789722222222222,
                },
            ),
            Point(
                Coord {
                    lat: 42.27305555555555,
                    lng: 8.788611111111111,
                },
            ),
            Point(
                Coord {
                    lat: 42.27222222222222,
                    lng: 8.788333333333334,
                },
            ),
            Point(
                Coord {
                    lat: 42.27166666666667,
                    lng: 8.787777777777778,
                },
            ),
            Point(
                Coord {
                    lat: 42.27111111111111,
                    lng: 8.7875,
                },
            ),
            Point(
                Coord {
                    lat: 42.27055555555555,
                    lng: 8.786944444444444,
                },
            ),
            Point(
                Coord {
                    lat: 42.269999999999996,
                    lng: 8.786666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 42.26916666666666,
                    lng: 8.786666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 42.2675,
                    lng: 8.7875,
                },
            ),
            Point(
                Coord {
                    lat: 42.26694444444444,
                    lng: 8.789444444444444,
                },
            ),
            Point(
                Coord {
                    lat: 42.266666666666666,
                    lng: 8.79,
                },
            ),
            Point(
                Coord {
                    lat: 42.26583333333333,
                    lng: 8.790555555555555,
                },
            ),
            Point(
                Coord {
                    lat: 42.265277777777776,
                    lng: 8.790833333333333,
                },
            ),
            Point(
                Coord {
                    lat: 42.263888888888886,
                    lng: 8.790555555555555,
                },
            ),
            Point(
                Coord {
                    lat: 42.26305555555555,
                    lng: 8.789444444444444,
                },
            ),
            Point(
                Coord {
                    lat: 42.2625,
                    lng: 8.790833333333333,
                },
            ),
            Point(
                Coord {
                    lat: 42.263333333333335,
                    lng: 8.7925,
                },
            ),
            Point(
                Coord {
                    lat: 42.26444444444444,
                    lng: 8.792222222222222,
                },
            ),
            Point(
                Coord {
                    lat: 42.265277777777776,
                    lng: 8.793055555555556,
                },
            ),
            Point(
                Coord {
                    lat: 42.264722222222225,
                    lng: 8.793888888888889,
                },
            ),
            Point(
                Coord {
                    lat: 42.26416666666667,
                    lng: 8.795555555555556,
                },
            ),
            Point(
                Coord {
                    lat: 42.26416666666667,
                    lng: 8.797777777777778,
                },
            ),
            Point(
                Coord {
                    lat: 42.263888888888886,
                    lng: 8.79888888888889,
                },
            ),
            Point(
                Coord {
                    lat: 42.263333333333335,
                    lng: 8.799444444444445,
                },
            ),
            Point(
                Coord {
                    lat: 42.2625,
                    lng: 8.799722222222222,
                },
            ),
            Point(
                Coord {
                    lat: 42.26222222222222,
                    lng: 8.8,
                },
            ),
            Point(
                Coord {
                    lat: 42.26222222222222,
                    lng: 8.801111111111112,
                },
            ),
            Point(
                Coord {
                    lat: 42.2625,
                    lng: 8.801944444444445,
                },
            ),
            Point(
                Coord {
                    lat: 42.2625,
                    lng: 8.8025,
                },
            ),
            Point(
                Coord {
                    lat: 42.26277777777778,
                    lng: 8.803333333333335,
                },
            ),
            Point(
                Coord {
                    lat: 42.2625,
                    lng: 8.80388888888889,
                },
            ),
            Point(
                Coord {
                    lat: 42.26277777777778,
                    lng: 8.804444444444446,
                },
            ),
            Point(
                Coord {
                    lat: 42.26277777777778,
                    lng: 8.805000000000001,
                },
            ),
            Point(
                Coord {
                    lat: 42.2625,
                    lng: 8.805555555555557,
                },
            ),
            Point(
                Coord {
                    lat: 42.26222222222222,
                    lng: 8.80638888888889,
                },
            ),
            Point(
                Coord {
                    lat: 42.2625,
                    lng: 8.806944444444445,
                },
            ),
            Point(
                Coord {
                    lat: 42.2625,
                    lng: 8.808055555555557,
                },
            ),
            Point(
                Coord {
                    lat: 42.26222222222222,
                    lng: 8.80888888888889,
                },
            ),
            Point(
                Coord {
                    lat: 42.26222222222222,
                    lng: 8.810833333333335,
                },
            ),
            Point(
                Coord {
                    lat: 42.26166666666666,
                    lng: 8.811666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 42.26166666666666,
                    lng: 8.813055555555556,
                },
            ),
            Point(
                Coord {
                    lat: 42.26083333333333,
                    lng: 8.814722222222223,
                },
            ),
            Point(
                Coord {
                    lat: 42.26166666666666,
                    lng: 8.815555555555557,
                },
            ),
            Point(
                Coord {
                    lat: 42.261944444444445,
                    lng: 8.816666666666666,
                },
            ),
            Point(
                Coord {
                    lat: 42.26277777777778,
                    lng: 8.818055555555555,
                },
            ),
            Point(
                Coord {
                    lat: 42.26361111111111,
                    lng: 8.81861111111111,
                },
            ),
            Point(
                Coord {
                    lat: 42.265,
                    lng: 8.82,
                },
            ),
            Point(
                Coord {
                    lat: 42.265277777777776,
                    lng: 8.820833333333333,
                },
            ),
            Point(
                Coord {
                    lat: 42.26638888888889,
                    lng: 8.821944444444444,
                },
            ),
            Point(
                Coord {
                    lat: 42.26722222222222,
                    lng: 8.821944444444444,
                },
            ),
            Point(
                Coord {
                    lat: 42.26777777777777,
                    lng: 8.821666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 42.26916666666666,
                    lng: 8.821666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 42.27027777777778,
                    lng: 8.821388888888889,
                },
            ),
        ],
    },
    frequency: None,
    call_sign: None,
    transponder_code: None,
}
